### PR TITLE
Dart - update generator to support null safety

### PR DESCRIPTION
**Pull request by agilob**
Targeting 6.x as there's plenty of breaking changes here

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob @ahmednfwela

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache [hunk 1]
```diff
@@ -4,12 +4,12 @@ class {{{classname}}} {
   const {{{classname}}}._(this.value);
 
   /// The underlying value of this enum member.
-  final {{{dataType}}} value;
+  final {{{dataType}}}? value;
```
Comment (Mike278) [inline modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache:7]: can this nullability depend on whether `enumVars` contains `null`?

File: modules/openapi-generator/src/main/resources/dart2/api_helper.mustache [hunk 2]
```diff
@@ -67,39 +67,27 @@ String parameterToString(dynamic value) {
...
-        return DateTime.tryParse(value);
+        return DateTime.parse(value);
       }
     }
     if (millis != null) {
       return DateTime.fromMillisecondsSinceEpoch(millis, isUtc: true);
     }
   }
-  return null;
+  return DateTime(1970);
```
Comment (Mike278) [inline modules/openapi-generator/src/main/resources/dart2/api_helper.mustache:92]: this should probably still return null and callsites should be updated to check `required`/`nullable` and add `!` if necessary

**Comment by devjakobsen** [comment]
Looking forward to the null safety code @agilob - good job. Can you add some task for the stuff that is missing? then I'll try to help out :)

**Comment by devjakobsen** [comment]
I tried out the changes - here's what I did.

1. ran command: `git clone https://github.com/agilob/openapi-generator.git`
2. ran command:  `./run-in-docker.sh mvn package.`
3. ran command: 
`./run-in-docker.sh generate \
    -i "openapi_spec.yaml" \
    --skip-validate-spec \
    --global-property models,modelDocs=false,modelTests=false \
    --global-property apis,apiTests=false,apiDocs=false \
    --global-property supportingFiles \
    -g dart \
    -o /gen/out/dartgenerated`

- It generated the dart files with success.
- It still had the `// @dart=2.0` header so you can't use it in a null safety project
- If you manual remove the header it had compile errors.

Could you point me in a direction how to solve these?

**Comment by ahmednfwela** [comment]
@devjakobsen did you checkout the correct branch?

**Comment by devjakobsen** [comment]
> @devjakobsen did you checkout the correct branch?
No, silly me. It works like a charm now :) Good job.
Do you know how to set the auth token on the Api Client? Haven't looked into it yet.

**Comment by devjakobsen** [comment]
@agilob or anyone else added a PR for optional parameter so you can pass a token to the API client if you use another auth method. https://github.com/agilob/openapi-generator/pull/17 - can you have a look?

**Comment by agilob** [comment]
Thanks for the PR @devjakobsen, I'm on holiday this week and early next week. I'll be happy to accept further fixes if you have time, if not I'll look into it again soon

**Comment by devjakobsen** [comment]
@agilob I was missing the autogenerated files from build on previous PR. If you merge https://github.com/agilob/openapi-generator/pull/19 and https://github.com/agilob/openapi-generator/pull/18. Then the build error "Some checks were not successful" should hopefully go away. Enjoy your holiday :)

File: modules/openapi-generator/src/main/resources/dart2/api_helper.mustache (hunk 2 above)
Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/api_helper.mustache]: Thanks for spotting this, it was on my list, but I had to miss it, fixed.

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache (hunk 1 above)
Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache:7]: do you mean "if any enumVar is null then this is nullable`?

**Comment by agilob** [comment]
@ahmednfwela @devjakobsen I updated branch with one more fix, if you could please test it or review it would be great

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache [hunk 3]
```diff
@@ -75,10 +75,10 @@ class {{{classname}}} {
   static {{{classname}}} fromJson(Map<String, dynamic> json) => {{{classname}}}(
   {{#vars}}
     {{#isDateTime}}
-        {{{name}}}: mapDateTime(json, r'{{{baseName}}}', '{{{pattern}}}'),
+        {{{name}}}: mapDateTime(json, r'{{{baseName}}}', '{{{pattern}}}'){{^isNullable}}!{{/isNullable}},
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: shouldn't you check for `isRequired` as well ?

Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: as in `if (required and notNullable) OR (notRequired and nullable)`?

Comment (ahmednfwela) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: well, that's what I was discussing in https://github.com/OpenAPITools/openapi-generator/issues/10972 
if the date can be `notNullable` and `notRequired`, should we assume it's nullable or not?

Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: I saw that discussion, but I dont have a standpoint there. If we (or you ;)) decide on a route to go, we should migrate all dart generators to make them compatible with each other. Until this is the case, I'm happy to go whichever you guys decide as a temporary solution. only nullable checks is my way as it's "dart-null" and "openapi-null", it is the same semantic, despite it has different meanings.

Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: Im leaving this as is, but here's a snippet for future if needed `{{#required}}{{^isNullable}}!{{/isNullable}}{{/required}}{{^required}}{{#isNullable}}!{{/isNullable}}{{/required}}`

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache [hunk 4]
```diff
@@ -1,16 +1,16 @@
 {{>header}}
 {{>part_of}}
 class ApiClient {
-  ApiClient({this.basePath = '{{{basePath}}}'}) {
+  ApiClient({this.basePath = '{{{basePath}}}', HttpBearerAuth? httpBearerAuth}) {
```
Comment (SimonIT) [inline modules/openapi-generator/src/main/resources/dart2/api_client.mustache:4]: Maybe add `{{#isBasicBearer}}` also here? It's kinda weird to have an constructor parameter without ausage

**Comment by SimonIT** [comment]
The line https://github.com/agilob/openapi-generator/blob/9518ee721c6a8b1164c482e45f0d3b473292cf0a/modules/openapi-generator/src/main/resources/dart2/api.mustache#L191 gives me a warning of
```
'Future.value' should not be called with a null argument for the non-nullable type argument 'NearestResponse'  Try adding a non-null argument.
```
This
```dart
return Future<{{{returnType}}}?>.value();
```
would resolve it

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache (hunk 4 above)
Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache:4]: nice

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache (hunk 1 above)
Comment (Mike278) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_enum.mustache:7]: yeah thats what I mean

I'm guessing `if any enumVar is null` would need to be exposed on the java side?

**Comment by agilob** [comment]
No more comments?

**Comment by noordawod** [comment]
https://github.com/OpenAPITools/openapi-generator/pull/11149

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache (hunk 3 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:78]: I used the following in my pr, imho it's a better one because it takes into consideration the `defaultValue`:

`{{#required}}{{^isNullable}}!{{/isNullable}}{{/required}}{{^required}}{{#defaultValue}} ?? {{{.}}}{{/defaultValue}}{{/required}}`

The pr: https://github.com/OpenAPITools/openapi-generator/pull/11149

**Comment by agilob** [comment]
Failure unrelated:

>Error:  Plugin org.jacoco:jacoco-maven-plugin:0.8.7 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.jacoco:jacoco-maven-plugin:jar:0.8.7: Could not transfer artifact org.jacoco:jacoco-maven-plugin:pom:0.8.7 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/jacoco/jacoco-maven-plugin/0.8.7/jacoco-maven-plugin-0.8.7.pom: Connection timed out (Read failed) -> [Help 1]

**Comment by agilob** [comment]
Ooops, I merged master instead 6.x, force pushing correct merge commit...

**Comment by agilob** [comment]
thanks @noordawod :+1:

File: modules/openapi-generator/src/main/resources/dart2/api.mustache [hunk 5]
```diff
@@ -49,24 +49,13 @@ class {{{classname}}} {
...
-
-  {{/hasParams}}
+  Future<Response> {{{nickname}}}WithHttpInfo({{#allParams}}{{#required}}{{{dataType}}} {{{paramName}}},{{^-last}} {{/-last}}{{/required}}{{/allParams}}{{#hasOptionalParams}}{ {{#allParams}}{{^required}}{{{dataType}}}? {{{paramName}}},{{^-last}} {{/-last}}{{/required}}{{/allParams}} }{{/hasOptionalParams}}) async {
     // ignore: prefer_const_declarations
     final path = r'{{{path}}}'{{#pathParams}}
       .replaceAll({{=<% %>=}}'{<% baseName %>}'<%={{ }}=%>, {{{paramName}}}{{^isString}}.toString(){{/isString}}){{/pathParams}};
 
     // ignore: prefer_final_locals
-    Object postBody{{#bodyParam}} = {{{paramName}}}{{/bodyParam}};
+    Object{{^bodyParam}}?{{/bodyParam}} postBody{{#bodyParam}} = {{{paramName}}}{{/bodyParam}};
```
Comment (SimonIT) [inline modules/openapi-generator/src/main/resources/dart2/api.mustache:58]: This creates for me some errors
```dart
  Future<Response> apiV3WorkPackagesIdPatchWithHttpInfo(int id, { bool? notify, WorkPackage? workPackage, }) async {
    // ignore: prefer_const_declarations
    final path = r'/api/v3/work_packages/{id}'
      .replaceAll('{id}', id.toString());

    // ignore: prefer_final_locals
    Object postBody = workPackage;
```
`workPackage` is nullable but `postBody` is not
spec: https://raw.githubusercontent.com/wbt-solutions/openproject_dart_sdk/master/openProject_swagger.yaml

File: samples/openapi3/client/petstore/dart2/petstore_client_lib_fake/lib/auth/http_basic_auth.dart [hunk 6]
```diff
@@ -11,12 +11,14 @@
 part of openapi.api;
 
 class HttpBasicAuth implements Authentication {
+  HttpBasicAuth(this.username, this.password);
```
Comment (SimonIT) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib_fake/lib/auth/http_basic_auth.dart:14]: The problem with the constructor is that it is used in the `ApiClient` constructor without arguments

File: modules/openapi-generator/src/main/resources/dart2/api.mustache (hunk 5 above)
Comment (agilob) [reply modules/openapi-generator/src/main/resources/dart2/api.mustache:58]: @noordawod could you please take a look at this?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api.mustache:58]: Should be an easy fix. How would you like me to post a fix? A pr on top of your pr, or give me temporary access?

File: samples/openapi3/client/petstore/dart2/petstore_client_lib_fake/lib/auth/http_basic_auth.dart (hunk 6 above)
Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib_fake/lib/auth/http_basic_auth.dart:14]: This used to be different actually, but copied from @agilob's pr. I can revert back to my version.

Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib_fake/lib/auth/http_basic_auth.dart]: :heavy_check_mark:

File: modules/openapi-generator/src/main/resources/dart2/api.mustache (hunk 5 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api.mustache]: :heavy_check_mark:

File: modules/openapi-generator/src/main/resources/dart2/auth/http_basic_auth.mustache [hunk 7]
```diff
@@ -1,14 +1,14 @@
 {{>header}}
 {{>part_of}}
 class HttpBasicAuth implements Authentication {
-  HttpBasicAuth(this.username, this.password);
+  HttpBasicAuth({this.username, this.password});
 
-  String username;
-  String password;
+  String? username;
```
Comment (agilob) [inline modules/openapi-generator/src/main/resources/dart2/auth/http_basic_auth.mustache:6]: I dont think making these nullable make any sense, they can't be null in HTTP spec, and we could get rid of null check below. Empty string also isn't a valid credential, you can't submit `agilob:` or `:password`. I think we should force users to provide non-nullable credentials and not do any checks on them.

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart [hunk 8]
```diff
@@ -11,14 +11,14 @@
 part of openapi.api;
 
 class HttpBasicAuth implements Authentication {
-  HttpBasicAuth(this.username, this.password);
+  HttpBasicAuth({this.username, this.password});
```
Comment (agilob) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:14]: same change here, this shouldnt be nullable or optional

File: modules/openapi-generator/src/main/resources/dart2/auth/http_basic_auth.mustache (hunk 7 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/auth/http_basic_auth.mustache:6]: The code has a call to `HttpBasicAuth()` which will produce empty strings. How do we handle that?

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart (hunk 8 above)
Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:14]: In this case, we could keep the code as-is (but make both properties non-nullable), but instead change `applyToParams()` to conditionally include a header if both are non-empty, what do you think?

Comment (agilob) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:14]: yea, sounds good. In general, not only this case, I think generated code shouldn't have null checks and we shouldn't decide what to put in place of a null object when we can move this responsibility on users of the API. They should decide if they want empty string, empty map/list etc. It removes responsibility and future proofs the API. If for some reason, it becomes necessary to change it, it will be easier to make this nullable, than not-nullable - wouldn't be a breaking change.

Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:14]: Sounds good. I can make the necessary changes in around 2 hours, going for a long walk :)

**Comment by noordawod** [comment]
@agilob the code now defaults to empty strings, no nulls anymore in `auth` directory.

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart [hunk 9]
```diff
@@ -11,14 +11,19 @@
...
-  String? username;
-  String? password;
+  String username;
+  String password;
 
   @override
   void applyToParams(List<QueryParam> queryParams, Map<String, String> headerParams) {
-    final credentials = '${username ?? ''}:${password ?? ''}';
-    headerParams['Authorization'] = 'Basic ${base64.encode(utf8.encode(credentials))}';
+    username = username.trim();
```
Comment (agilob) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:21]: why the trim?

Comment (agilob) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:21]: if a username of password ends with ` ` space user won't be able to login as wrong password will be sent to a server, user generated API has not way to fix it without modifying generated code, which then would be overriden in the next generation.

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_bearer_auth.dart [hunk 10]
```diff
@@ -21,19 +21,31 @@ class HttpBearerAuth implements Authentication {
...
+      accessToken = _accessToken;
     } else if (_accessToken is HttpBearerAuthProvider) {
-      headerParams['Authorization'] = 'Bearer ${_accessToken()}';
+      accessToken = _accessToken!();
     } else {
-      throw ArgumentError('Type of Bearer accessToken should be a String or a String Function().');
+      return;
+    }
+
+    accessToken = accessToken.trim();
```
Comment (agilob) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_bearer_auth.dart:45]: same here, this can't be trimmed as we would be modifying user input.

Comment (agilob) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_bearer_auth.dart:45]: If a user wants to send value -1, we don't want to change it to 0 or 1 just because it's user provided negative number and we don't like negative numbers? openapi should not modify or edit any values user provided.Same reason as before, we can't convert null to empty string just because we would like it so. User wants to send null, null it is, user wants to send `\n\n\r\t` (EOL, EOL windows, tab) as username, we send it this way. if user provides `{password="password"}` as password, this clearly looks like JSON, but we have no business decoding it. generated code must act as a dumb actor and not modify or edit provided data.

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart (hunk 9 above)
Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart:21]: Got it, I'll fix then and upload new code.

Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_basic_auth.dart]: :heavy_check_mark:

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_bearer_auth.dart (hunk 10 above)
Comment (noordawod) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/auth/http_bearer_auth.dart]: :heavy_check_mark:

**Comment by agilob** [comment]
@jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob @ahmednfwela @noordawod any last reviews? can I get green ticks on this PR please?

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache [hunk 11]
```diff
@@ -180,37 +225,56 @@ class {{{classname}}} {
...
+    return result.toList(growable: growable);
+  }
 
   static Map<String, {{{classname}}}> mapFromJson(dynamic json) {
     final map = <String, {{{classname}}}>{};
     if (json is Map && json.isNotEmpty) {
-      json
-        .cast<String, dynamic>()
-        .forEach((key, dynamic value) => map[key] = {{{classname}}}.fromJson(value));
+      json = json.cast<String, dynamic>(); // ignore: parameter_assignments
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:244]: can you move this to top of the file? since this gets repeated with every function

**Review by ahmednfwela — APPROVED** [review]

**Comment by ahmednfwela** [comment]
note: the java changes also affect other generators

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java [hunk 12]
```diff
@@ -64,7 +64,7 @@
     public String defaultValue;
     public String arrayModelType;
     public boolean isAlias; // Is this effectively an alias of another simple type
-    public boolean isString, isInteger, isLong, isNumber, isNumeric, isFloat, isDouble, isDate, isDateTime, isShort, isUnboundedInteger, isBoolean;
+    public boolean isString, isInteger, isLong, isNumber, isNumeric, isFloat, isDouble, isDate, isDateTime, isShort, isUnboundedInteger, isPrimitiveType, isBoolean;
```
Comment (kuhnroyal) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: Why this change? Doesn't seem to be set/used.
Also how can a model be primitive?

**Review by kuhnroyal — CHANGES_REQUESTED** [review]
Not looking at all the Dart changes, its just too much.

**Review by noordawod — APPROVED** [review]
Amazing work :star_struck:

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java (hunk 12 above)
Comment (ahmednfwela) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: I would expect this to be an auto property combining all known OAS3 primitive types, e.g.
```java
isString || isInteger || isLong || isNumber || isNumeric || isFloat || isDouble || isDate || isDateTime || isShort || isUnboundedInteger ...
```

**Comment by noordawod** [comment]
> Not looking at all the Dart changes, its just too much.

Yeah, it took almost a day's work to painstakinly test, retest and reshape the changes, hopefully with amicable results for the public. But for me it still works in real-life apps and now it's null safe.

**Review by sbu-WBT — APPROVED** [review]
LGTM

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java (hunk 12 above)
Comment (agilob) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: while this doesn't look to be part of Dart2 functionality, it makes sense to add this, as concept of `isPrimitiveType` is known in other generators like in cpp and rust. This change adds functionality of `isPrimitiveType` check to _all_ generators. Going to keep it here.

**Comment by agilob** [comment]
@sbu-WBT pls give green tick

**Comment by agilob** [comment]
@wing328 can we have this in to 6.x please! :tada:

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java (hunk 12 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: But it is not being set anywhere, this is missleading

Comment (ahmednfwela) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: I agree that this should be a separate PR and get discussed between other generator implementers

**Comment by agilob** [comment]
@wing328 pls :eyes:

**Comment by wing328** [comment]
Just merged. Sorry for the delay.

**Comment by agilob** [comment]
well done everyone :)

**Comment by noordawod** [comment]
Is there an estimate when v6 will be merged to `master`?

**Comment by agilob** [comment]
This month most likely, with 6.0 release

**Comment by fatihozdil** [comment]
auto generated readme points to use the get method(apiVVersionPersonnelGet) but this method does not return anything 
image

image

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache [hunk 13]
```diff
@@ -135,38 +180,38 @@ class {{{classname}}} {
...
             {{#isMap}}
-        {{{name}}}: mapCastOfType<String, {{{items.datatype}}}>(json, r'{{{baseName}}}'),
+        {{{name}}}: mapCastOfType<String, {{{items.datatype}}}>(json, r'{{{baseName}}}'){{#required}}{{^isNullable}}!{{/isNullable}}{{/required}}{{^required}}{{#defaultValue}} ?? {{{.}}}{{/defaultValue}}{{/required}},
             {{/isMap}}
             {{^isMap}}
               {{#isNumber}}
         {{{name}}}: json[r'{{{baseName}}}'] == null
-          ? null
-          : {{{datatypeWithEnum}}}.parse(json[r'{{{baseName}}}'].toString()),
+            ? {{#defaultValue}}{{{.}}}{{/defaultValue}}{{^defaultValue}}null{{/defaultValue}}
```
Comment (eseidel) [inline modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:206]: As far as I can tell, this line isn't quite right.  It doesn't handle required num types, which will be (correctly) generated as non-nullable, but then this line tries to set them to null when they don't have a default value which won't compile.

https://raw.githubusercontent.com/SpaceTradersAPI/api-docs/b52734a2cdb63e796174310ebac89f5009df3b9f/reference/SpaceTraders.json

is an example of a spec which exhibits this, specifically:
https://raw.githubusercontent.com/SpaceTradersAPI/api-docs/b52734a2cdb63e796174310ebac89f5009df3b9f/models/JumpGate.json

Comment (ahmednfwela) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:206]: @eseidel can you post this in a separate issue?
After these 2 PRs https://github.com/OpenAPITools/openapi-generator/pull/14346 , https://github.com/OpenAPITools/openapi-generator/issues/15477 are merged
we can now unify all efforts into one generator, which means solving a lot of bugs
